src/cpp/redis.cpp

```diff
@@ -49,62 +49,36 @@ Redis::~Redis()
         delete this->_redis;
 }
 
-CommandReply Redis::run(Command& cmd)
-{
-    Command::iterator cmd_fields_start = cmd.begin();
-    Command::iterator cmd_fields_end = cmd.end();
-    CommandReply reply;
-
-    int n_trials = 100;
-
-    while (n_trials > 0) {
+CommandReply Redis::run(SingleKeyCommand& cmd){
+    return this->_run(cmd);
+}
 
-        try {
-            reply = this->_redis->command(cmd_fields_start, cmd_fields_end);
+CommandReply Redis::run(MultiKeyCommand& cmd){
+    return this->_run(cmd);
+}
 
-            if(reply.has_error()==0)
-                return reply;
-            n_trials = 0;
-        }
-        catch (sw::redis::IoError &e) {
-            n_trials--;
-            std::this_thread::sleep_for(std::chrono::seconds(2));
-        }
-        catch (sw::redis::ClosedError &e) {
-            n_trials--;
-            std::this_thread::sleep_for(std::chrono::seconds(2));
-        }
-        catch (std::exception& e) {
-            throw std::runtime_error(e.what());
-        }
-        catch (...) {
-            throw std::runtime_error("A non-standard exception "\
-                                     "encountered during command " +
-                                     cmd.first_field() +
-                                     " execution. ");
-        }
-    }
+CommandReply Redis::run(CompoundCommand& cmd){
+    return this->_run(cmd);
+}
 
-    if (n_trials == 0) {
-        if(reply.has_error()>0)
-            reply.print_reply_error();
-        throw std::runtime_error("Redis failed to execute command: " +
-                                 cmd.first_field());
-    }
+CommandReply Redis::run(AddressAtCommand& cmd){
+    return this->_run(cmd);
+}
 
-    return reply;
+CommandReply Redis::run(AddressAnyCommand& cmd){
+    return this->_run(cmd);
 }
 
-CommandReply Redis::run(CommandList& cmds)
+std::vector<CommandReply> Redis::run(CommandList& cmds)
 {
+    std::vector<CommandReply> replies;
     CommandList::iterator cmd = cmds.begin();
     CommandList::iterator cmd_end = cmds.end();
-    CommandReply reply;
     while(cmd != cmd_end) {
-        reply = this->run(**cmd);
+        replies.push_back(dynamic_cast<Command*>(*cmd)->run_me(this));
```

Comment:
With this implementation, if any command fails badly enough to throw an exception, none of the subsequent ones will be attempted. I think this is what you want, but if you feel otherwise you should capture the result into a temporary `CommandReply` and put the call into a try/catch block.